### ISSUE TITLE
bson.Setter not working on slices

### ISSUE THREAD

**Issue by niemeyer** [go-mgo/mgo#69]
Reported in the mailing list:

https://groups.google.com/d/msg/mgo-users/78Vs4nUm4Fo/ZQw79UXSoEAJ

**Comment by inlined** [comment]
Was about to file an issue that slice unmarshalling doesn't work period; Setter or no. I was able to marshal an []int32 but failed to unmarshal because it was an unsupported type. bsonspec.org clarifies this is marshalled into a map {'0': firstVal, '1': secondVal} and binary inspection confirms this. I'm currently working around this limitation with

```
var workaround map[string]int32
if err := bson.Unmarshal(bytes, &workaround); err != nil {
  return nil, err
}
var vals []int32
for _, elem := range workaround {
  vals = append(vals, elem)
}
return vals, nil
```

Despite the fact that the bytes were created by successfully encoding a []int32.
Should this be folded into the current issue or should I file a new one?

**Comment by niemeyer** [comment]
A BSON array is a list of elements, not a map, and as such it needs to be unmarshaled into a Go slice, or into a JavaScript Array if you are using the console. If you are trying to unmarshal a BSON array into a Go map, that will not work, and is correct behavior.

All of these issues are also unrelated to this ticket. If you have code demonstrating the above not working, please file an independent issue, or come to the mailing list so we can talk about it.

**Comment by niemeyer** [comment]
Issue fixed in 7a49434.
